### PR TITLE
31 feature dynamic model fetching caching selection and metadata

### PR DESCRIPTION
**Pull request by WiegerWolf**
image

## Summary by CodeRabbit

- **New Features**
  - Introduced dynamic model management with asynchronous loading, selection, and switching across CLI and GUI.
  - Added UI feedback for model loading states and updated model lists.
  - Enhanced message display to indicate the generating model.
  - Persisted user-selected model and model metadata between sessions.
  - Added AI prompt templates to guide feature planning and stepwise implementation.

- **Improvements**
  - Improved database schema with model metadata, message timestamps, and migration support.
  - Enhanced error handling and retry logic for API calls and model loading.
  - Added cross-platform support for database storage paths.
  - Refined tool message validation and cleanup logic.

- **Bug Fixes**
  - Fixed orphaned tool message cleanup and added JSON validation for tool messages.

- **Chores**
  - Added default model ID and API endpoint constants to configuration.
  - Refactored interfaces and code to integrate model management and UI updates.

### CODE REVIEW

**Comment by WiegerWolf** [comment]
Regarding the earlier comment about `persistence_manager.h`: This file was indeed redundant and has now been removed. Thanks for the catch!

File: chat_client.h [hunk 1]
```diff
@@ -19,7 +24,25 @@ class ChatClient {
...
+    
+    // --- Model Selection State (Part III GUI Changes) ---
+    std::string active_model_id;
+    // const std::string default_model_id = "phi3:mini"; // Replaced by DEFAULT_MODEL_ID from config.h
+    // --- End Model Selection State ---
+
+    // For model initialization and management
+    std::atomic<bool> models_loading{false}; // For UI feedback
+    std::future<void> model_load_future;     // For asynchronous loading
+
```
[the remark this answers is not in the document]
Comment (WiegerWolf) [reply chat_client.h:37]: Added a destructor to `ChatClient` that waits for `model_load_future` to complete if it's still valid. This should prevent the potential UB. Thanks for pointing this out!

File: chat_client.h [hunk 2]
```diff
@@ -19,7 +24,25 @@ class ChatClient {
...
     UserInterface& ui; // Add reference to the UI
     std::string api_base = "https://openrouter.ai/api/v1/chat/completions";
-    std::string model_name = "openai/gpt-4.1-nano";
+    // std::string model_name = "openai/gpt-4.1-nano"; // Replaced by active_model_id
+    
+    // --- Model Selection State (Part III GUI Changes) ---
+    std::string active_model_id;
+    // const std::string default_model_id = "phi3:mini"; // Replaced by DEFAULT_MODEL_ID from config.h
+    // --- End Model Selection State ---
+
```
[the remark this answers is not in the document]
Comment (WiegerWolf) [reply chat_client.h:33]: Thanks for the suggestion! I've confirmed that `active_model_id` is indeed initialized with `DEFAULT_MODEL_ID` (defined in [`config.h.in:7`](config.h.in:7)) in the `ChatClient` constructor at [`chat_client.cpp:41`](chat_client.cpp:41). So, this concern should already be addressed.

File: gui_interface/gui_interface.h [hunk 3]
```diff
@@ -104,6 +126,16 @@ void setInitialFontSize(float size); // Added for persistence
...
 private:
+    PersistenceManager& db_manager_ref; // Reference to PersistenceManager
+
+    // --- Model Selection State (Part III GUI Changes / Updated Part V) ---
+    std::string current_selected_model_id_in_ui; // Renamed for clarity
+    // const std::string default_model_id = "phi3:mini"; // This is now DEFAULT_MODEL_ID from config.h
+    std::vector<ModelEntry> available_models_for_ui; // Cache for UI display
+    std::atomic<bool> models_are_loading_in_ui{false}; // For UI feedback
+    mutable std::mutex models_ui_mutex; // To protect available_models_for_ui and current_selected_model_id_in_ui
+    // --- End Model Selection State ---
```
[the remark this answers is not in the document]
Comment (WiegerWolf) [reply gui_interface/gui_interface.h:137]: Thanks for the review. I've looked into this, and the atomic flag (`models_are_loading_in_ui`) and the mutex (`models_ui_mutex`) serve distinct and complementary purposes here:
*   The **atomic flag** provides a lightweight, lock-free mechanism for frequent UI status checks of the model loading *process* (e.g., to show/hide a loading spinner) without the overhead of acquiring a mutex.
*   The **mutex** ensures thread-safe access and modification of the *shared data structures* (`available_models_for_ui` vector and `current_selected_model_id_in_ui` string), as operations on these are not inherently atomic.
This separation prevents race conditions on the data while allowing for efficient status polling. Therefore, both are necessary and not redundant in this context.

File: gui_interface/gui_interface.cpp [hunk 4]
```diff
@@ -624,3 +637,111 @@ void GuiInterface::processFontRebuildRequest() {
...
+        std::lock_guard<std::mutex> lock(models_ui_mutex);
+        this->current_selected_model_id_in_ui = model_id;
+    } // Mutex released before DB operation
+    try {
+        db_manager_ref.saveSetting("selected_model_id", model_id);
+    } catch (const std::exception& e) {
+        std::cerr << "Error saving selected model ID: " << e.what() << std::endl;
+        // displayError("Failed to save model selection preference."); // Can be noisy
+    }
+}
```
[the remark this answers is not in the document]
Comment (WiegerWolf) [reply gui_interface/gui_interface.cpp:678]: Thanks for raising this concern. I've reviewed the lock ordering between `models_ui_mutex` and the database mutex. The analysis indicates that a deadlock risk from inconsistent ordering is unlikely here.
The established pattern is to acquire `models_ui_mutex` first, and then, if database operations are needed, they proceed, which involves SQLite's internal mutex. We haven't found any code paths that attempt to acquire these locks in the reverse order. Database operations are generally self-contained and don't call back to lock `models_ui_mutex` while a database transaction is active.
A more detailed look at this has been documented in `deadlock_analysis_gui_db.md`.

File: gui_interface/gui_interface.cpp [hunk 5]
```diff
@@ -624,3 +637,111 @@ void GuiInterface::processFontRebuildRequest() {
...
+            if (model_data && !model_data->name.empty()) {
+                name = model_data->name;
+            }
+        } catch (...) { /* ignore, use constructed name */ }
+        fallback_entries.push_back({current_id, name});
+        return fallback_entries;
+    }
+    return available_models_for_ui;
+}
+
```
[the remark this answers is not in the document]
Comment (WiegerWolf) [reply gui_interface/gui_interface.cpp:665]: Thanks for the feedback on `getAvailableModelsForUI`. I've reviewed its implementation. The method primarily serves the model list from an internal cache (`available_models_for_ui`), which is populated by `updateModelsList`. It doesn't fetch the full list from the database on every call.
There is a fallback where it might fetch details for a single model directly from the DB if the main cache is empty, but this is not its primary mode of operation for retrieving the entire list.
Your broader point about ensuring efficient data loading and avoiding unnecessary DB lookups for the UI is well taken. We'll keep an eye on how `ChatClient` fetches the model data that populates this cache and the frequency of `updateModelsList` calls to ensure overall performance.

File: gui_interface/gui_interface.cpp [hunk 6]
```diff
@@ -39,9 +40,12 @@ static void glfw_error_callback(int error, const char* description) {
...
 static bool imgui_init_done = false;
 
-GuiInterface::GuiInterface() {
+GuiInterface::GuiInterface(PersistenceManager& db_manager) : db_manager_ref(db_manager) {
     // Initialize the input buffer
     input_buf[0] = '\0';
+    // current_selected_model_id_in_ui will be initialized in initialize()
+    // available_models_for_ui will be populated by updateModelsList
+    // models_are_loading_in_ui defaults to false (atomic)
 }
```
[the remark this answers is not in the document]
Comment (WiegerWolf) [reply gui_interface/gui_interface.cpp:49]: Thanks for the suggestion regarding the `window` pointer initialization. I've checked the code, and the `window` member is already initialized to `nullptr` via in-class member initialization in `gui_interface.h:109` (i.e., `GLFWwindow* window = nullptr;`). This ensures it's safely initialized before any potential use. So, this concern should already be covered.

File: gui_interface/gui_interface.cpp [hunk 7]
```diff
@@ -624,3 +637,111 @@ void GuiInterface::processFontRebuildRequest() {
...
+        
+        // Persist this newly determined selection
+        try {
+             db_manager_ref.saveSetting("selected_model_id", current_selected_model_id_in_ui);
+        } catch (const std::exception& e) {
+            std::cerr << "Error saving fallback selected model ID in updateModelsList: " << e.what() << std::endl;
+        }
+    }
+    // The main_gui.cpp render loop will use getAvailableModelsForUI() to refresh the dropdown.
+}
```
[the remark this answers is not in the document]
Comment (WiegerWolf) [reply gui_interface/gui_interface.cpp:746]: Refactored `updateModelsList` to move the `saveSetting` database call outside the `models_ui_mutex` lock. This should help minimize the critical section. Thanks for the catch!

File: database.h [hunk 8]
```diff
@@ -3,39 +3,80 @@
...
+#include "model_types.h" // For ModelData struct
+struct sqlite3; // Forward declaration for SQLite database handle
 
 struct Message {
     std::string role;
     std::string content;
     int id = 0;
     std::string timestamp; // Added timestamp field
+    std::string model_id; // <-- New field (changed from optional)
 };
```
[the remark this answers is not in the document]
Comment (WiegerWolf) [reply database.h:15]: Implemented changes to make `Message::model_id` backward compatible using `std::optional<std::string>`. Legacy assistant messages will now get an 'UNKNOWN_LEGACY_MODEL_ID', and all consuming code has been updated. Thanks for highlighting this!

File: chat_client.cpp [hunk 9]
```diff
@@ -35,151 +35,391 @@ ChatClient::ChatClient(UserInterface& ui_ref, PersistenceManager& db_ref) :
...
     auto headers_guard = std::unique_ptr<struct curl_slist, decltype(&curl_slist_free_all)>{nullptr, curl_slist_free_all};
 
     headers = curl_slist_append(headers, ("Authorization: Bearer " + api_key).c_str());
-    headers_guard.reset(headers); // Transfer ownership to guard
     headers = curl_slist_append(headers, "Content-Type: application/json");
-    headers = curl_slist_append(headers, "HTTP-Referer: https://llm-cli.tsatsin.com");
-    headers = curl_slist_append(headers, "X-Title: LLM-cli");
+    headers = curl_slist_append(headers, "HTTP-Referer: https://llm-cli.tsatsin.com"); // Replace with your actual site
+    headers = curl_slist_append(headers, "X-Title: LLM-cli"); // Replace with your actual app name
+    headers_guard.reset(headers);
```
[the remark this answers is not in the document]
Comment (WiegerWolf) [reply chat_client.cpp:309]: Refactored the `goto try_api_call;` in `makeApiCall` to use a loop for the retry logic.
Regarding the `goto cleanup_and_return;` and `Json::Value/Reader` mentioned in the comment: I searched the file and did not find this specific goto or Jsoncpp types.

File: docs/step-2.md [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+        ```cpp
+        // Ensure model_init_thread is joinable before starting a new one if constructor can be called multiple times
+        // (though typically it's once per ChatClient instance)
+        if (model_init_thread.joinable()) {
+            model_init_thread.join(); 
+        }
+        model_init_thread = std::thread(&ChatClient::initializeModels, this);
+        model_init_thread.detach(); // Detach to run in background
+        ```
+    *   *Correction*: Detaching might not be ideal if the main application relies on models being loaded soon. Consider if a join with timeout or a more robust signaling mechanism is needed for the UI to know when it's safe to operate. For now, per requirements, "run in a separate thread" implies background. Detach is simplest for plan.
```
[the remark this answers is not in the document]
Comment (WiegerWolf) [reply docs/step-2.md:126]: Updated the documentation in [`docs/step-2.md:126`](docs/step-2.md:126) to accurately describe the model initialization process, which uses `std::async` and manages the future's lifetime correctly (including in the destructor). The previous `std::thread::detach` example was misleading and has been corrected. The underlying code concern is already addressed.

File: database.cpp [hunk 11]
```diff
@@ -6,97 +6,198 @@
...
-            throw std::runtime_error("Failed to get HOME directory environment variable.");
+// Constructor for the implementation class
+PersistenceManager::Impl::Impl() : db(nullptr) { // Initialize db pointer
+    // Construct the database path in the user's home directory
+    const char* home_dir = std::getenv("HOME");
+    if (!home_dir) {
+        throw std::runtime_error("Failed to get HOME directory environment variable.");
+    }
+    std::string path = std::string(home_dir) + "/.llm-cli-chat.db";
+
```
[the remark this answers is not in the document]
Comment (WiegerWolf) [reply database.cpp:17]: Implemented a cross-platform solution for database path construction using `std::filesystem` and a helper function (`get_home_directory_path`) to find the home directory. The database will now be stored at `~/.llm-cli/llm_chat_history.db` with a fallback to the current directory if the home directory cannot be determined or path creation fails. Thanks for the suggestion!

File: database.cpp [hunk 12]
```diff
@@ -6,97 +6,198 @@
...
+        sqlite3_bind_text(stmt, 3, msg.model_id.c_str(), -1, SQLITE_STATIC);
+    } else {
+        sqlite3_bind_null(stmt, 3);
+    }
+
+    // Execute the prepared statement.
+    if(sqlite3_step(stmt) != SQLITE_DONE) {
+        throw std::runtime_error("Insert failed: " + std::string(sqlite3_errmsg(db)));
+    }
+}
```
[the remark this answers is not in the document]
Comment (WiegerWolf) [reply database.cpp:149]: Applied RAII (using `std::unique_ptr` with a custom deleter) to the `sqlite3_stmt` used for the `PRAGMA table_info` check in the `PersistenceManager` constructor. This ensures the statement is properly finalized. Thanks for the catch!

File: database.cpp [hunk 13]
```diff
@@ -6,97 +6,198 @@
...
+    char* err = nullptr;
+    if(sqlite3_exec(db, sql, nullptr, nullptr, &err) != SQLITE_OK) {
+        std::string msg = "SQL error executing '";
+        msg += sql;
+        msg += "': ";
+        msg += err;
+        sqlite3_free(err); // Free the error message allocated by SQLite
+        throw std::runtime_error(msg);
+    }
+}
```
[the remark this answers is not in the document]
Comment (WiegerWolf) [reply database.cpp]: Added null checks for the error message pointer in the `sqlite3_exec` error handling within `PersistenceManager::Impl::exec` methods. This will prevent potential null dereferences. Thanks for spotting this!
